### PR TITLE
Configurable RAM to VRAM percentage setting for integrated devices

### PR DESCRIPTION
**Pull request by scorpion81**
This shall allow to specify a certain percentage of RAM to use as VRAM on integrated / shared devices.
Ideally this can help tuning the garbage collector to prevent out of memory crashes.
On the other hand a too aggressive memory limitation may indeed affect rendering negatively
I did experience this on the steam deck for example. Graphical glitches may be caused in low memory
situations.
Note: there is still no proper detection yet for being actually on an integrated device. It is just a rough idea on maybe how to possibly tune the garbage collection a bit. This should affect both OpenGL and Vulkan renderers.
The default values at the moment are VRAM percentage setting is enabled and is set to 25% of the total RAM. Note on steam deck the total RAM is reported as 11.5 GB with UMA Buffer size at 4 GB, and 14.5 GB with UMA Buffer size at 1 GB.
Using for example 40% for VRAM seemed for now(!) to fix even the recall animation issue, but this all needs further testing.
Once good hardcoded defaults are found, this may be obsolete again.

### CODE REVIEW

File: src/common/settings.cpp [hunk 1]
```diff
@@ -250,4 +255,23 @@ void RestoreGlobalState(bool is_powered_on) {
     values.motion_enabled.SetGlobal(true);
 }
 
+// this function only makes sense with integrated devices
+
+u64 RAM_Percent_to_Byte(u8 percent) {
```
Comment (liamwhite) [inline src/common/settings.cpp:260]: RamPercentageToBytes or RAMPercentageToBytes for yuzu's code style

Comment (scorpion81) [reply src/common/settings.cpp]: Ok, did rename the function now to "RamPercentageToBytes", and fixed a lot of formatting stuff :smile:

**Comment by Alleop5** [comment]
Would this pull request make it possible for me to launch TOTK on a iGPU with 256 MBs of VRAM?

**Comment by Squall-Leonhart** [comment]
> Would this pull request make it possible for me to launch TOTK on a iGPU with 256 MBs of VRAM?

No.

**Comment by scorpion81** [comment]
> Would this pull request make it possible for me to launch TOTK on a iGPU with 256 MBs of VRAM?

Well I am only able to test with my steam deck, setting the UMA Framebuffer size down to 256 MB in BIOS. And it seemed to work normally here. 
This basically means that the minimal amount of ram being usable as vram is 256 MB, and that more of the RAM is being used as VRAM if necessary. 
But this I think only works if your system / iGPU is capable to grab more RAM automatically for using as VRAM. 
This pr cannot do magic and add more memory. But it can try to adapt the garbage collection base memory amount, determines when the garbage collection starts roughly spoken. 
In case of Vulkan, if this setting is disabled, this limit is about 4 GB. With this setting it might be tuned a little bit differently, basically overriding that 4 GB limit to a percentage of the current RAM (ranging from 10% - 90%). But this must be tested on different devices with different amounts of RAM, how this limit may affect the memory usage and / or performance.
In case of OpenGL, it tries to take this setting into account too. But for the OpenGL case I didnt find a way yet to detect whether the device is integrated or not, so this setting would apply to all GPUs. Means it should perhaps be disabled, in case something stupid happens on dedicated GPUs with OpenGL.
So it is the long version of the "No", as in this only works if your system can automatically reclaim more VRAM from RAM.

**Comment by FearlessTobi** [comment]
We have a huge amount of settings in the advanced graphics tab already... can't we try to determine the amount of vram required automatically?

**Comment by scorpion81** [comment]
> We have a huge amount of settings in the advanced graphics tab already... can't we try to determine the amount of vram required automatically?

Indeed this should be either a good hardcoded value or a value based on the current total amount of RAM and available amount of RAM. In fact those UI settings were added solely to test the behavior of different percentages of VRAM as input for the garbage collector calculations, without recompiling the whole thing over and over again in order to test new values. And the root cause of those VRAM tests was #10493 .
But it turned out this issue starts appearing inconsistently with the same percentage of VRAM set. The only duct-tape fix was to increase the percentage so the issue wont appear, but avoiding to increase it too much so the RAM didnt fully fill up. 
So i think this only is treating the symptoms, not the cause of the particular issue.
Beyond this issue, i am not really sure whether it would be useful to be able to influence the garbage collector from "outside".
To determine a good value we would need to take into account whether we have Vulkan or OpenGL active (since the gc collections differ between them), whether we have an integrated gpu or not (since this only makes sense for iGPUs) and how much of the shared RAM we would like to use. Besides one hardcoded value we could use different values maybe for 4 GB, 8 GB, 16 GB , 32 GB RAM and so on available. 
From testing a value between 33% and 37% of the available RAM turned out to be a good compromise. For example, on the steam deck with 1 GB UMA Framebuffer set,  about ~14.5 GB of free RAM are being reported to yuzu, and applying that percentage yielded values about 5 - 5.5-ish GB. Going higher caused the RAM slowly filling up entirely and going lower cause the issue to be reproducibly appear. But this applied only to vulkan, since I didnt manage to get this properly working for OpenGL at all.
So as conclusion, I would suggest to maybe increase the 4_GB hard cap to about 5_GB or 5_GB + 512_MB or 6_GB for 16 GB RAM systems. 
The exposure to the UI could actually lead to more potential to misconfigure yuzu as well, and was mainly for testing purposes.
So I am closing this PR now, but feel free to re-open it in case it may turn out to be useful elsewhere as well.

**Pull request closed without merging**